[git-webkit] Prevent global PR search while using `git-webkit pr`

youssefsoli opened the pull request:
#### 2aeaf88debf7c143bf65d6969893c2b090dd4740
<pre>
[git-webkit] Prevent global PR search while using `git-webkit pr`
<a href="https://bugs.webkit.org/show_bug.cgi?id=242801">https://bugs.webkit.org/show_bug.cgi?id=242801</a>

Reviewed by Jonathan Bedard.

Currently, betweeen the pre-PR checks and pushing to the remote fork steps
within `git-webkit pr`, the tool attempts to find if the current branch is
already being used for a pull request on GitHub by using the `/repos/{owner}/{repo}/pulls`
API. The search can take in a `base` parameter to refine the search for
pulls matching a certain `user:branchname` combination.

If the `:` delimiter isn&apos;t present in the base parameter, GitHub will search
every PR that has been made within a repository with a limit of 100 results
per query. The tool incorrectly uses the `base` name as just the branch
without a user or organization before the delimiter alongside using
`currentuser:branchname`.

Since pagination is enabled by default, this would make existing PRs be
found with only 1 request to GitHub, while PRs that don&apos;t exist will take
(at the moment of writing this message) ~24 requests containing every PR
made in the repository. This results in the noticeable lag between the
style checker and the remote push and also can cause someone to hit the
API limit fairly quickly.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.find):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLandGitHub):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(test_update):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/squash_unittest.py:
(TestSquash.test_github_with_previous_history):
(TestSquash.test_github_without_previous_history):
(TestSquash.test_github_two_step):

Canonical link: <a href="https://commits.webkit.org/252508@main">https://commits.webkit.org/252508@main</a>
</pre>

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
```diff
@@ -75,9 +75,12 @@ def find(self, opened=True, head=None, base=None):
+            elif user and head:
+                heads.append('{}:{}'.format(user, head))
+
             for qhead in heads:
```
Doesn't look like this should be a list any more, right? There doesn't seem to be a way to have multiple `heads` now.

youssefsoli committed:
[git-webkit] Prevent global PR search while using `git-webkit pr`
https://bugs.webkit.org/show_bug.cgi?id=242801

Reviewed by Jonathan Bedard.

Currently, betweeen the pre-PR checks and pushing to the remote fork steps
within `git-webkit pr`, the tool attempts to find if the current branch is
already being used for a pull request on GitHub by using the `/repos/{owner}/{repo}/pulls`
API. The search can take in a `base` parameter to refine the search for
pulls matching a certain `user:branchname` combination.

If the `:` delimiter isn't present in the base parameter, GitHub will search
every PR that has been made within a repository with a limit of 100 results
per query. The tool incorrectly uses the `base` name as just the branch
without a user or organization before the delimiter alongside using
`currentuser:branchname`.

Since pagination is enabled by default, this would make existing PRs be
found with only 1 request to GitHub, while PRs that don't exist will take
(at the moment of writing this message) ~24 requests containing every PR
made in the repository. This results in the noticeable lag between the
style checker and the remote push and also can cause someone to hit the
API limit fairly quickly.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.find):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLandGitHub):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(test_update):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/squash_unittest.py:
(TestSquash.test_github_with_previous_history):
(TestSquash.test_github_without_previous_history):
(TestSquash.test_github_two_step):

Canonical link: https://commits.webkit.org/252508@main
